soc: dts: drivers: add SoCs for amebadplus series

zjian-zhang opened the pull request:
The [amebadplus](https://www.realmcu.com/en/Home/Product/add965ea-d661-4a63-9514-d18b6912f8ab) microcontroller (MCU) is an advanced system-on-chip (SoC) featuring an Arm® Cortex®-M55 and Cortex®-M23, large Flash and SRAM memories, and supports dual-band Wi-Fi 4(2.4GHz + 5GHz) and BLE5.0 specifications.

The initial support for the board includes GPIO, PINCTRL and Loguart(serial).

zjian-zhang committed:
boards: add rtl872xda_evb board

add initial version of rtl872xda_evb board

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

nordicjm commented on boards/realtek/rtl872xda_evb/rtl872xda_evb_defconfig:
```diff
@@ -0,0 +1,54 @@
+
+# enable rcc
+CONFIG_CLOCK_CONTROL=y
+CONFIG_PINCTRL=y
```
Kconfigs select this

nordicjm replied:
Not addressed

zjian-zhang replied:
Fixed, I misunderstood and just changed line 8.

nordicjm commented on boards/realtek/rtl872xda_evb/rtl872xda_evb_defconfig:
```diff
@@ -0,0 +1,54 @@
+CONFIG_CLOCK_CONTROL=y
+CONFIG_PINCTRL=y
+
+# enable uart
```
Comments should be e.g. `# Enable UART`

nordicjm commented on boards/realtek/rtl872xda_evb/rtl872xda_evb_defconfig:
```diff
@@ -0,0 +1,54 @@
+CONFIG_SHELL=y
+CONFIG_LOG=y
+
+# enable trng
+CONFIG_ENTROPY_GENERATOR=y
+CONFIG_DMA=y
+CONFIG_FLASH=y
+# CONFIG_SPI=y
+CONFIG_I2C=y
+CONFIG_WATCHDOG=y
+
+# CONFIG_GPIO=y
+CONFIG_RTC=y
+CONFIG_ADC=y
+CONFIG_LED_STRIP=y
+CONFIG_COUNTER=y
+CONFIG_PWM=y
+CONFIG_PWM_CAPTURE=y
+CONFIG_I2S=y
+CONFIG_AUDIO=y
+CONFIG_AUDIO_DMIC=y
+CONFIG_CRYPTO=y
```
No to any of this

nordicjm commented on soc/realtek/ameba/amebadplus/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	default "amebadplus" if SOC_SERIES_AMEBADPLUS
+
+config SOC_AMEBADPLUS
+	bool "Real-M300 on AmebaDplus"
```
no selection on this

nordicjm commented on soc/realtek/ameba/amebadplus/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	select SOC_SERIES_AMEBADPLUS
+
+config SOC
+	default "SOC_AMEBADPLUS" if SOC_AMEBADPLUS
```
no SOC_ on text here

nordicjm commented on soc/realtek/ameba/amebadplus/Kconfig:
```diff
@@ -0,0 +1,16 @@
+	select ARMV8_M_DSP
+	select CPU_HAS_ARM_MPU
+	select ARM_TRUSTZONE_M
+	# select SHARED_MULTI_HEAP
+	# select SYSTICK_USE_EXTERNAL_CLOCK
```
get rid of commented lines

nordicjm reviewed (changes requested):
Please read and follow the contribution guidelines https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribute-guidelines we do not shove everything into a single commit for multiple areas, they must be distinct

jfischer-no commented on drivers/usb/udc/udc_ameba.c:
```diff
@@ -0,0 +1,897 @@
+/*
```
Please open a separate PR to add the USB device controller driver and include information about the controller type. But first check if the controller type is already supported, for designs implemented by different vendors I would ask for a native driver implementation. At first glance it looks like this SoC has a DWC2 compatible controller, if so you need to use the udc_dwc2 driver.

zjian-zhang replied:
I have remove usb related code, and will open a separate PR in next commit to submit usb.

zjian-zhang replied:
done!

zjian-zhang commented on west.yml:
```diff
@@ -23,12 +23,26 @@ manifest:
+      path: modules/hal/realtek/ameba
+      revision: master
+      groups:
+        - hal
```
new modification has been submitted...

henrikbrixandersen reviewed (dismissed):
It's not just the USB support. You need to split this up into multiple PRs, one for adding the basic support including GPIO and UART - the rest needs to go into separate PRs, as the different areas require reviews from different maintainers.

zjian-zhang committed:
soc: add realtek amebadplus SOC integration

Add initial version of Amebadplus Soc integration

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

zjian-zhang committed:
drivers: pinctrl: add amebadplus pin controller driver

add amebadplus pin controller driver

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

zjian-zhang committed:
drivers: serial: add amebadplus loguart driver

loguart driver for amebadplus

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

zjian-zhang committed:
west.yml: Add Realtek HAL as a new HAL module

Realtek HAL (Hardware Abstraction Layer) provides
a low level peripheral configuration function.

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

zjian-zhang committed:
drivers: gpio: add amebadplus gpio driver

GPIO driver for amebadplus

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

jfischer-no commented on soc/realtek/ameba/amebadplus/ameba_rom_symbol_acut_s.ld:
```diff
@@ -0,0 +1,800 @@
+    PROVIDE(USB_EPStartXfer = 0xf86d);
+    PROVIDE(USB_ReadPacket = 0xfa01);
+    PROVIDE(USB_EPSetStall = 0xfa21);
+    PROVIDE(USB_EPClearStall = 0xfa5d);
+    PROVIDE(USB_StopDevice = 0xfa99);
+    PROVIDE(USB_SetDevAddress = 0xfae5);
+    PROVIDE(USB_DevConnect = 0xfb05);
+    PROVIDE(USB_DevDisconnect = 0xfb25);
+    PROVIDE(USB_ReadInterrupts = 0xfb45);
+    PROVIDE(USB_ReadDevAllOutEpInterrupt = 0xfb55);
+    PROVIDE(USB_ReadDevAllInEpInterrupt = 0xfb65);
+    PROVIDE(USB_ReadDevOutEPInterrupt = 0xfb75);
+    PROVIDE(USB_ReadDevInEPInterrupt = 0xfb8d);
+    PROVIDE(USB_ClearInterrupts = 0xfba5);
+    PROVIDE(USB_GetMode = 0xfbb5);
+    PROVIDE(USB_ActivateSetup = 0xfbc5);
+    PROVIDE(USB_EP0_OutStart = 0xfbfd);
+    PROVIDE(USBD_IsAttached = 0xff31);
+    PROVIDE(USBD_Process = 0xff3d);
+    PROVIDE(USBD_Init = 0xffb1);
+    PROVIDE(USBD_DeInit = 0xffed);
+    PROVIDE(USB_PCD_Init = 0xfff5);
+    PROVIDE(USB_PCD_Start = 0x101ed);
+    PROVIDE(USB_PCD_Stop = 0x10215);
+    PROVIDE(USB_PCD_DeInit = 0x10251);
+    PROVIDE(USB_PCD_DevConnect = 0x10291);
+    PROVIDE(USB_PCD_DevDisconnect = 0x102b5);
+    PROVIDE(USB_PCD_SetAddress = 0x102d9);
+    PROVIDE(USB_PCD_EP_Open = 0x10301);
+    PROVIDE(USB_PCD_EP_Close = 0x10379);
+    PROVIDE(USB_PCD_EP_Receive = 0x103cd);
+    PROVIDE(USB_PCD_EP_GetRxCount = 0x1041d);
+    PROVIDE(USB_PCD_EP_Transmit = 0x1042d);
+    PROVIDE(USB_PCD_EP_SetStall = 0x10481);
+    PROVIDE(USB_PCD_EP_ClrStall = 0x104e1);
+    PROVIDE(USB_PCD_EP_Flush = 0x1053d);
+    PROVIDE(USB_PCD_GetState = 0x10579);
+    PROVIDE(USB_PCD_SetTxFiFo = 0x10581);
+    PROVIDE(USB_PCD_SetRxFiFo = 0x105d9);
+    PROVIDE(USB_PCD_IRQHandler = 0x105e5);
```
What is it and why is it there?

zjian-zhang replied:
amebadplus has a rom that contains some driver code, so it is the address of function in rom. shall I move it to hal_realtek repository?

jfischer-no replied:
Why is it there and how do you think it could be used in the Zephyr RTOS? Did you understand my previous comment about the USB driver?

zjian-zhang replied:
1. amebadplus support trustzone, so cpu will run in secure world and non-secure world, here are two files in soc/realtek/ameba/amebadplus folder, ameba_rom_symbol_acut_s.ld and ameba_rom_symbol_acut.ld, ameba_rom_symbol_acut_s.ld is used in bootloader(like mucboot) and a loader for download image when mass production, there are all run in secure world; and ameba_rom_symbol_acut.ld is used in application image(such as freertos or zephyr), but for now we run zephyr in secure world, so you can see some code like usb/xmoderm, which is used for a loader in mass production. when we porting Trusted Firmware-M done, zephyr will run in non-secure world, and **the ameba_rom_symbol_acut.ld will be used, so the usb function in ameba_rom_symbol_acut_s.ld could not be used in the zephyr RTOS.**
2. your previous comment about the USB driver, here is reply from our team:
Dear Sir
There are two reasons why we can't use the DWC2 driver.
1) Our chips use a shared FIFO for all non-periodic (CTRL/BULK/INTR) IN endpoints, therefore, based on the DWC2 driver in the current zephyr SDK(3.7.0), our chips can not work because the DWC2 "only dedicated TX FIFO mode is supported".
We use the Shared NPTxFIFO, so we need to deal with the EPMIS interrupt and intoken learning queue, which are very important for us to complete the USB transfer.
2) the endpoint direction is limited in our chips, not all endpoints can support two-way transmission, for example, Endpoint 1 is only used for USB In transfers, Endpoint 2 is only used for USB Out transfers.

Thanks !

jfischer-no replied:
>     1. Our chips use a shared FIFO for all non-periodic (CTRL/BULK/INTR) IN endpoints, therefore, based on the DWC2 driver in the current zephyr SDK(3.7.0), our chips can not work because the DWC2 "only dedicated TX FIFO mode is supported".
>        We use the Shared NPTxFIFO, so we need to deal with the EPMIS interrupt and intoken learning queue, which are very important for us to complete the USB transfer.
> 

You probably do not understand how to work on a common infrastructure. You are welcome to add support for the shared FIFO mode. Seriously, there will not be another driver for the same controller in Zephyr tree.

>     2. the endpoint direction is limited in our chips, not all endpoints can support two-way transmission, for example, Endpoint 1 is only used for USB In transfers, Endpoint 2 is only used for USB Out transfers.

I do not see any problems with that.

zjian-zhang replied:
I will discuss this with our team. but the symbol PROVIDE(USB_xxxx = 0xxxxx); is just in the ROM of the amebadplus asic which is used by the loader for mass production, and not used by zephyr. and this file was copied from our [github:](https://github.com/Ameba-AIoT/ameba-rtos/blob/master/amebadplus_gcc_project/project_km4/asdk/ld/ameba_rom_symbol_acut_s.ld) with no modification.
 Anyway, I have removed it for this PR.

zjian-zhang replied:
our team agree with what you said.

zjian-zhang replied:
Our team agrees with what you said and add the modification when commit usb dirver. thanks..

zjian-zhang commented:
Please help to check again if there are any problems. I have other drivers to submit, which were separated from of this PR. thanks.

nordicjm commented on boards/realtek/rtl872xda_evb/doc/index.rst:
```diff
@@ -0,0 +1,119 @@
+Supported Features
+==================
+
+Current Zephyr's RTL872XDA-EVB board supports the following features:
```
needs proper board name as used for west build with \`\` around it, referred to as board target

zjian-zhang replied:
Done! but `Supported Features` has been updated to `.. zephyr:board-supported-hw::`

nordicjm commented on boards/realtek/rtl872xda_evb/doc/index.rst:
```diff
@@ -0,0 +1,119 @@
+Prerequisites
+-------------
+
+Realtek HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
+
+.. code-block:: console
+
+   west blobs fetch hal_realtek
+
+.. note::
+
+   It is recommended running the command above after :file:`west update`.
```
??? there is no such blobs in zephyr, so this can go

zjian-zhang replied:
fixed, it was added for next PR/issue, we will add blob to hal_realtek git repository.

nordicjm replied:
Well no you won't add a blob file directly to that repo, it needs to be hosted elsewhere as per https://docs.zephyrproject.org/latest/contribute/bin_blobs.html

zjian-zhang replied:
yes, i am waiting this pr done, and edit a issue to summit it.

nordicjm commented on boards/realtek/rtl872xda_evb/doc/rtl872xda_evb.png:
convert to webp, then put through https://tinypng.com/

nordicjm commented on boards/realtek/rtl872xda_evb/rtl872xda_evb-pinctrl.dtsi:
```diff
@@ -0,0 +1,20 @@
+&pinctrl {
+	compatible = "realtek,ameba-pinctrl";
+
+	/*PA30 is SWD_CLK, PA31 is SWD_DAT(both pull-up internally)*/
```
spaces after `/*` and before `*/`

nordicjm reviewed (changes requested):
Not looking part first commit if previous comments in that commit have not been addressed

nordicjm commented on boards/realtek/rtl872xda_evb/doc/index.rst:
```diff
@@ -0,0 +1,102 @@
+Supported Features
+==================
+
+Current Zephyr's ``rtl872xda_evb`` board supports the following features:
```
board target

nordicjm commented on boards/realtek/rtl872xda_evb/doc/rtl872xda_evb.webp:
Can you reduce resolution of this? We don't need a 1400 pixel wide image for board documentation

zjian-zhang replied:
Done!

nordicjm commented on soc/realtek/ameba/soc.yml:
```diff
@@ -0,0 +1,6 @@
+family:
+- name: realtek_ameba
+  series:
+  - name: amebadplus
+    socs:
+    - name: amebadplus
```
Is this really true, you have a line of products called `amebadplus` and one specific model on that line is called `amebadplus`? Series is not required, if it is not needed then drop the series

nordicjm commented on soc/realtek/ameba/common/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2024 Realtek Semiconductor Corp.
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_include_directories(.)
```
Huh? You have a common folder with just a cmake file and include this directory, which is empty?

nordicjm commented on drivers/gpio/Kconfig.ameba:
```diff
@@ -0,0 +1,16 @@
+	depends on GPIO_AMEBA
+	default y
+	help
+	  When Enable GPIO Interrupt Debounce, the external signal can be debounced to remove any spurious glitches that are less than one period(about 32us) of the external debouncing clock.
```
maximum length of 100 characters, wrap lines (1x tab, followed by 2x spaces for each help text line)

zjian-zhang replied:
Done~

nordicjm reviewed (commented):
Commits must be placed in order, base soc files, then drivers, then board files, you can't add board files first that reference things that do not exist.

Onto the hal module, what's going on with that? There's files that include freertos headers, there's some configuration file that defines e.g. lfs is enabled, this is not how things are done in zephyr, there's files that adds fake posix functions? What is any of this any why is it there?

nordicjm commented:
Also dts entries in the .dtsi must explicity be marked as `status = "disabled";`

zjian-zhang committed:
dts: arm: introduce amebadplus SOC Devicetree

add initial version of devicetree for amebadplus SOC.
amebadplus devicetree file is main platform dtsi file, which should
be included from board dts (e.g rtl872xda_evb.dts)

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

zjian-zhang commented:
> Commits must be placed in order, base soc files, then drivers, then board files, you can't add board files first that reference things that do not exist.
> 
> Onto the hal module, what's going on with that? There's files that include freertos headers, there's some configuration file that defines e.g. lfs is enabled, this is not how things are done in zephyr, there's files that adds fake posix functions? What is any of this any why is it there?

1. The commit has been ordered. We refer to other PR first, but your guidance is more conform to [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribute-guidelines).
2. freertos should be commented out, configuration file abort lfs has been modified. fake posix functions have been removed, there are used for zero warning in our sdk when building with no OS.

zjian-zhang commented:
> Also dts entries in the .dtsi must explicity be marked as `status = "disabled";`

is there all entry shall be marked as `status = "disabled";` ? I have check other company's .dtsi, and leave the serial's status okay. wating for your guidance.

nordicjm commented:
> > Also dts entries in the .dtsi must explicity be marked as `status = "disabled";`
> 
> is there all entry shall be marked as `status = "disabled";` ? I have check other company's .dtsi, and leave the serial's status okay. wating for your guidance.

Essential functionality e.g. clocks, regulators, the SoC timing source etc. should be enabled, then things that should be enabled in boards like UART etc. should be enabled in the board dts, everything else e.g. spi etc. should be disabled in SoC dtsi

zjian-zhang commented:
> > > Also dts entries in the .dtsi must explicity be marked as `status = "disabled";`
> > 
> > 
> > is there all entry shall be marked as `status = "disabled";` ? I have check other company's .dtsi, and leave the serial's status okay. wating for your guidance.
> 
> Essential functionality e.g. clocks, regulators, the SoC timing source etc. should be enabled, then things that should be enabled in boards like UART etc. should be enabled in the board dts, everything else e.g. spi etc. should be disabled in SoC dtsi

Thanks, Based on your feedback, I've made several changes and pass the CI tests. Could you please re-review when you have a moment?

zjian-zhang commented:
Please help to check again if there are any problems. thanks.

nordicjm commented:
I am not happy with the hal cmake code, you are following a "throw everything and see what sticks" approach by compiling every .c file then letting the linker drop what is not needed https://github.com/zjian-zhang/hal_realtek/commit/da530b55b9b6ab019d4d4b4f9c289316013dc7ee#diff-da88b28506b76806121c844bc9143929cd342413a3b62089f0a309ab1e3356ab
This is a problem because your boards will be built in CI, which already has a high load, and you're adding further strain to it by building things that are just going to be binned. These files should only be included when they are needed (i.e. selected with Kconfig). You can see how vendors do this like so https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrfx/CMakeLists.txt#L117

zjian-zhang commented:
> I am not happy with the hal cmake code, you are following a "throw everything and see what sticks" approach by compiling every .c file then letting the linker drop what is not needed [zjian-zhang/hal_realtek@da530b5#diff-da88b28506b76806121c844bc9143929cd342413a3b62089f0a309ab1e3356ab](https://github.com/zjian-zhang/hal_realtek/commit/da530b55b9b6ab019d4d4b4f9c289316013dc7ee#diff-da88b28506b76806121c844bc9143929cd342413a3b62089f0a309ab1e3356ab) This is a problem because your boards will be built in CI, which already has a high load, and you're adding further strain to it by building things that are just going to be binned. These files should only be included when they are needed (i.e. selected with Kconfig). You can see how vendors do this like so https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrfx/CMakeLists.txt#L117
> 
> There's also lots of commented lines in there? Can you explain what the point of wrapping lots of C functions is? @keith-packard any comments on that?

I have modified CMakeLists.txt according to Kconfig and removed commented-out lines. Since the ROM code contains a large number of C functions, wrapping these functions can help save RAM/flash usage. Additionally, the rand wrapper is configured to use TRNG.

nordicjm commented:
> > I am not happy with the hal cmake code, you are following a "throw everything and see what sticks" approach by compiling every .c file then letting the linker drop what is not needed [zjian-zhang/hal_realtek@da530b5#diff-da88b28506b76806121c844bc9143929cd342413a3b62089f0a309ab1e3356ab](https://github.com/zjian-zhang/hal_realtek/commit/da530b55b9b6ab019d4d4b4f9c289316013dc7ee#diff-da88b28506b76806121c844bc9143929cd342413a3b62089f0a309ab1e3356ab) This is a problem because your boards will be built in CI, which already has a high load, and you're adding further strain to it by building things that are just going to be binned. These files should only be included when they are needed (i.e. selected with Kconfig). You can see how vendors do this like so https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrfx/CMakeLists.txt#L117
> > There's also lots of commented lines in there? Can you explain what the point of wrapping lots of C functions is? @keith-packard any comments on that?
> 
> I have modified CMakeLists.txt according to Kconfig and removed commented-out lines. Since the ROM code contains a large number of C functions, wrapping these functions can help save RAM/flash usage. Additionally, the rand wrapper is configured to use TRNG.

To me that's equivalent of using a binary blob. Thoughts @stephanosio @keith-packard ?

stephanosio commented on soc/realtek/ameba/common/os_wrapper/include/dlist.h:
```diff
@@ -0,0 +1,255 @@
+#ifndef __LIST_H
+#define __LIST_H
+
+/* This file is from Linux Kernel (include/linux/list.h)
+ * and modified by simply removing hardware prefetching of list items.
+ * Here by copyright, credits attributed to wherever they belong.
+ * Kulesh Shanmugasundaram (kulesh [squiggly] isis.poly.edu)
+ */
```
https://github.com/torvalds/linux/blob/6c52d4da1c742cd01a797a4d0a2d3c5a60dc9bfe/include/linux/list.h#L1

You can't just take GPL 2.0-licensed code and re-license it as Apache 2.0 ...

Also, why can't you use the existing list API in Zephyr?

zjian-zhang replied:
thanks, I have change to use the  existing list API in Zephyr.

zjian-zhang replied:
Done!

stephanosio commented:
> > I have modified CMakeLists.txt according to Kconfig and removed commented-out lines. Since the ROM code contains a large number of C functions, wrapping these functions can help save RAM/flash usage. Additionally, the rand wrapper is configured to use TRNG.
> 
> To me that's equivalent of using a binary blob. Thoughts @stephanosio @keith-packard ?

Well, as long as the "blob" is not part of the Zephyr image, I do not think this is a problem. It is not dissimilar to OS code calling proprietary system firmware code (e.g. BIOS, EFI) or even microcode ...

zjian-zhang commented:
Please help to check again if there are any problems. thanks.

rruuaanng commented on drivers/gpio/gpio_ameba.c:
```diff
@@ -0,0 +1,291 @@
+	/* port base address */
+	uint32_t base;
+	/* IO port */
+	int port_num;
```
Does it represent a `port` or the `number of ports`? It seems to be different from the comment.

```suggestion
	int port; // or pin
```

zjian-zhang replied:
done

zjian-zhang commented:
Please help to check again if there are any problems. thanks.

nordicjm commented on dts/arm/realtek/amebadplus/amebadplus.dtsi:
```diff
@@ -0,0 +1,101 @@
+			reg = <0x4100f000 0x100>;
+			interrupts = <27 0>;
+			current-speed = <1500000>;
+			status = "okay";
```
disabled

zjian-zhang replied:
Done!

nordicjm commented on dts/arm/realtek/amebadplus/amebadplus.dtsi:
```diff
@@ -0,0 +1,101 @@
+				compatible = "soc-nv-flash";
+				erase-block-size = <DT_SIZE_K(4)>;
+				write-block-size = <4>;
+				/* Flash size is specified in board dts */
```
how can a board set the size of flash on a soc?

zjian-zhang replied:
in 'boards/realtek/rtl872xda_evb/rtl872xda_evb.dts' we use
```
/* 4MB flash */
&flash0 {
	reg = <0x0e000020 DT_SIZE_M(4)>;
};
```

nordicjm replied:
But how? The flash is part of the SoC itself, how can it be defined at a board level?

zjian-zhang replied:
The flash is part of the board, not the SoC.

nordicjm commented on soc/realtek/ameba/amebadplus/soc.c:
```diff
@@ -0,0 +1,45 @@
+	return 0;
+}
+
+SYS_INIT(amebadplus_init, PRE_KERNEL_1, 0);
```
use soc hooks

zjian-zhang replied:
done

nordicjm commented on soc/realtek/ameba/common/os_wrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+zephyr_include_directories(include)
+
+# add_subdirectory(source)
```
can go?

zjian-zhang replied:
Done!

nordicjm commented on soc/realtek/ameba/common/os_wrapper/include/os_wrapper_memory.h:
```diff
@@ -0,0 +1,36 @@
+#define __OS_WRAPPER_MEMORY_H__
+
+/**
+ * @brief  Allocate memory from the heap. For FreeRTOS, map to pvPortMalloc
+ *         The buffer value is random
```
why are there comments about freertos?

zjian-zhang replied:
already removed

nordicjm commented on soc/realtek/ameba/common/os_wrapper/include/os_wrapper_mutex.h:
```diff
@@ -0,0 +1,68 @@
+ * handle.
+ * @retval
+ */
+int rtos_mutex_create(rtos_mutex_t *pp_handle);
```
@gmarull can you make heads or tails of this?

zjian-zhang replied:
Our driver requires semaphore services provided by the OS. By using this interface, our driver can support different operating systems.

nordicjm replied:
Why are the drivers not in zephyr in the `drivers` folder like every other vendor?

zjian-zhang replied:
Apologies for the misunderstanding. The code in the HAL repositories requires these APIs, and the review comment in the [HAL module request](https://github.com/zephyrproject-rtos/zephyr/issues/78084) stating ‘Zephyr glue code is NOT allowed in HAL repositories’ necessitates moving these APIs to the Zephyr repository.

nordicjm commented on boards/realtek/rtl872xda_evb/doc/index.rst:
```diff
@@ -0,0 +1,102 @@
+Supported Features
+==================
+
+Current Zephyr's board target ``rtl872xda_evb`` supports the following features:
```
```suggestion
Currently Zephyr's ``rtl872xda_evb`` board supports the following features:
```

nordicjm commented on boards/realtek/rtl872xda_evb/doc/index.rst:
```diff
@@ -0,0 +1,102 @@
+
+See the ApplicationNote chapter Image Tool from documentation links for more details.
+
+#. Environment Requirements: EX. WinXP, Win 7 or later, Microsoft .NET Framework 4.0.
```
so this device is completely unusable on linux or mac? `west flash` also cannot work?

zjian-zhang replied:
We currently do not support the ‘west flash’ command. In the future, we will provide a Python script to enable downloading across all platforms by loguart.

nordicjm commented on west.yml:
```diff
@@ -212,6 +214,12 @@ manifest:
       path: modules/hal/quicklogic
       groups:
         - hal
+    - name: hal_realtek
```
module commit needs to be before everything else, can't add code that uses a module before the module is present

zjian-zhang replied:
we have issue a HAL module request: [hal_realtek](https://github.com/zephyrproject-rtos/zephyr/issues/78084)

nordicjm replied:
it still needs to be the first commit here, you can amend the commit later when the HAL is approved and merged

zjian-zhang replied:
Thank you, I have adjusted the submission order, but the HAL has not been approved by the TSC yet. Could you help prioritize the review with the TSC?

zjian-zhang replied:
Done!

nordicjm commented on dts/arm/realtek/amebadplus/amebadplus.dtsi:
```diff
@@ -0,0 +1,100 @@
+		spic: flash-controller@40128000 {
+			compatible = "realtek,ameba-flash-controller";
+			reg = <0x40128000 0x200>;
+			/* interrupts = <51 0>; */
```
line to go?

zjian-zhang replied:
Done!

zjian-zhang commented:
Please help to check the issue. thanks a lot.

teburd commented on dts/bindings/pinctrl/realtek,ameba-pinctrl.yaml:
```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2024 Realtek Semiconductor Corp.
```
Lots of 2024 copyrights here, not a big deal to me but it is 2025

zjian-zhang replied:
That's a fair observation! The 2024 reflects when the project was initially created or files were generated. While some projects update yearly and you'll see newly added files dated 2025, others keep the original year. But I appreciate the attention to detail!

teburd reviewed (commented):
Looked it over, many things beyond my area of maintainership so I'm going to reassign, only one comment from me.

zjian-zhang committed:
MAINTAINERS: Add hal_realtek as a new HAL module

This commit adds maintainers for hal_realtek repository

Signed-off-by: zjian zhang <zjian_zhang@realsil.com.cn>

nordicjm commented on soc/realtek/ameba/common/os_wrapper/include/os_wrapper.h:
```diff
@@ -0,0 +1,47 @@
+#include "os_wrapper_semaphore.h"
+#include "os_wrapper_task.h"
+#include "os_wrapper_time.h"
+#include "os_wrapper_timer.h"
```
None of this is acceptable, anywhere. Your HAL is for the hardware, it should not be making threads or anything like that, if an **application** wants to make a thread, it will call the **zephyr** thread functions, it will not call some odd HAL-specific function which does something and redirects it

fabiobaltieri replied:
> By the way, Can the **DNM (manifest)** label be removed now? The PR is currently failing the `PR Metadata Check / Prevent Merging` due to the **DNM (manifest)** label.

The workflow adds it with every push because it detects a new module, just ignore it, we'll remove it before merging once it's ready, we can remove it now but then it'd just be added again on the next push.

nordicjm commented on soc/realtek/ameba/amebadplus/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+zephyr_linker_sources(SECTIONS boot_section.ld)
+set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/arch/arm/cortex_m/scripts/linker.ld CACHE INTERNAL "")
+
+set(PYTHON_SCRIPT_ARGS
```
in cmake UPPER CASE means GLOBAL, and lower case means local, fix, and move this inside of the if()

nordicjm commented on soc/realtek/ameba/common/os_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+#include <string.h>
+#include <stdint.h>
+
+#include "platform_autoconf.h"
```
actually whole file to go

nordicjm replied:
why would the build fail? There is no reference to these in this PR or the hal PR

nordicjm reviewed (changes requested):
~~SoC does not work without binary blobs as per https://github.com/zephyrproject-rtos/zephyr/pull/78036#discussion_r2592124755 therefore SoC is not acceptable in zephyr~~ clarified below

nordicjm commented on soc/realtek/ameba/amebadplus/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+    DEPENDS ${ZEPHYR_BINARY_DIR}/${KERNEL_ELF_NAME} ${ZEPHYR_HAL_REALTEK_MODULE_DIR}/ameba/${CONFIG_SOC_SERIES}/manifest.json5
+  )
+else()
+  message(WARNING "Missing Realtek blobs. Run: `west blobs fetch hal_realtek`")
```
OK please be very clear about that in future. And this warning needs removing

zjian-zhang replied:
Done!

nordicjm commented on soc/realtek/ameba/common/os_wrapper.h:
```diff
@@ -0,0 +1,16 @@
+#define __OS_WRAPPER_H__
+
+typedef void *rtos_sema_t;
+#define rtos_sema_create_binary(pp_handle)
```
file needs deleting

zjian-zhang replied:
Done!

zjian-zhang commented:
> It's not just the USB support. You need to split this up into multiple PRs, one for adding the basic support including GPIO and UART - the rest needs to go into separate PRs, as the different areas require reviews from different maintainers.

The earlier feedback on the first commit has been addressed. Thanks!

zjian-zhang commented on boards/realtek/rtl872xda_evb/doc/index.rst:
```diff
@@ -0,0 +1,102 @@
+| PINMUX     | on-chip    | pinctrl                             |
++------------+------------+-------------------------------------+
+| NVIC       | on-chip    | nested vector interrupt controller  |
++------------+------------+-------------------------------------+
```
Done!

zjian-zhang commented on drivers/gpio/Kconfig:
```diff
@@ -207,4 +207,6 @@ source "drivers/gpio/Kconfig.xlnx_ps"
 source "drivers/gpio/Kconfig.xmc4xxx"
 # zephyr-keep-sorted-stop
 
+source "drivers/gpio/Kconfig.ameba"
```
Done!

zjian-zhang commented on soc/realtek/ameba/common/os_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+ */
+#define RTOS_MAX_DELAY    0xFFFFFFFFUL
+#define RTOS_MAX_TIMEOUT  0xFFFFFFFFUL
+#define RTOS_TICK_RATE_HZ 1000UL
```
Done!

zjian-zhang commented on drivers/gpio/Kconfig:
```diff
@@ -207,4 +208,5 @@ source "drivers/gpio/Kconfig.xlnx_ps"
 source "drivers/gpio/Kconfig.xmc4xxx"
 # zephyr-keep-sorted-stop
 
+
```
Done!
